mountpoint-s3/src/inode.rs

```diff
@@ -540,12 +629,42 @@ impl ReaddirStreamState {
 
 impl ReaddirHandle {
     pub async fn next<OC: ObjectClient>(&self, client: &OC) -> Result<Option<LookedUp>, InodeError> {
-        while self.results.read().unwrap().is_empty() {
+        // We will start fetching new results when number of items in the remote results queue is empty
+        while self.remote_results.read().unwrap().is_empty() {
             let continuation_token = {
                 let mut next_token = self.next_continuation_token.lock().unwrap();
+
+                // populate local results before the first stream
+                if *next_token == ReaddirStreamState::NotStarted {
+                    let inode = self.inner.get(self.dir_ino)?;
+                    let kind_data = &inode.inner.sync.read().unwrap().kind_data;
+                    let local_files = match kind_data {
+                        InodeKindData::File { .. } => unreachable!("we know this is a directory"),
+                        InodeKindData::Directory {
+                            children: _,
+                            writing_children,
+                        } => writing_children.iter().map(|ino| {
+                            let inode = self.inner.get(*ino)?;
+                            let stat = inode.inner.sync.read().unwrap().stat.clone();
+                            Ok(LookedUp { inode, stat })
+                        }),
+                    };
+
+                    match local_files.collect::<Result<Vec<_>, _>>() {
+                        Ok(mut new_results) => {
+                            new_results.sort_by(|left, right| left.inode.name().cmp(right.inode.name()));
+                            self.local_results.write().unwrap().extend(new_results);
+                        }
+                        Err(e) => {
+                            error!(error=?e, "readdir failed");
+                            return Err(e);
+                        }
+                    }
+                }
+
                 if *next_token == ReaddirStreamState::Finished {
                     trace!(self=?self as *const _, "readdir finished");
-                    return Ok(None);
+                    return Ok(self.compare_and_get_next());
                 }
```

Comment:
Can we write a unit test (probably in this file) specifically to cover the cases around this logic?
1. Remote directory is totally empty
2. All local keys come after all remote keys

Reply:
Added these tests.